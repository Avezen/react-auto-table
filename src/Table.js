import React, {Component} from 'react';
import TruncatedString from "./TruncatedString";
import loader from "./assets/images/loader.svg";
import TableNavigation from "./TableNavigation";

class Table extends Component {

    render() {
        const { data, currentPage, setCurrentPage, pages, search, setSearch, sortBy, setSortBy, sortDirection, isFetching, error, colWidths, tableHead, currentlyOpenedTd, toggleTd } = this.props;

        return (
            <React.Fragment>
                <div
                    className={'table-nav--top'}
                >
                    <TableNavigation
                        setCurrentPage={setCurrentPage}
                        currentPage={currentPage}
                        pages={pages}
                    />
                    <input
                        className={'input'}
                        value={search}
                        type={'text'}
                        placeholder={'Search...'}
                        onChange={setSearch}
                    />
                </div>
                <table>
                    {colWidths && colWidths.length > 0 && (
                        <colgroup>
                            {colWidths.map((item, key) => (
                                <col
                                    key={key}
                                    width={item}
                                />
                            ))}
                        </colgroup>
                    )}

                    <thead>
                    <tr>
                        {tableHead.length > 0 && tableHead.map((item, key) => (
                            <th
                                key={key}
                                onClick={() => setSortBy(item.replace(/\s/g, ''))}
                            >
                                {item}
                                {item.replace(/\s/g, '') === sortBy && (
                                    sortDirection === 1 ? (
                                        <i className={'arrow down'}></i>
                                    ) : (
                                        <i className={'arrow up'}></i>
                                    )
                                )}
                            </th>
                        ))}
                        <th
                        >
                            Actions
                        </th>
                    </tr>
                    </thead>
                    <tbody>
                    {
                        !data ? (
                            <React.Fragment/>
                        ) : isFetching ? (
                            <tr>
                                <td colSpan={tableHead.length === 0 ? 1 : tableHead.length}>
                                    <div
                                        className={'fetch-info'}
                                    >
                                        <img
                                            src={loader}
                                            alt={'loader'}
                                            width={30}
                                        />
                                    </div>
                                </td>
                            </tr>
                        ) : error ? (
                            <tr>
                                <td colSpan={tableHead.length}>
                                    <div
                                        className={'fetch-info'}
                                    >
                                        Error while fetching data
                                    </div>
                                </td>
                            </tr>
                        ) : data.length > 0 ? (
                            data.map((item, key) => {
                                let dateString = (new Date(item.Deadline)).toLocaleString('pl-PL', {hour12: false});
                                dateString = dateString.substr(0, dateString.indexOf(','));

                                return (
                                    <tr key={key}>
                                        {Object.entries(item).map(([key, value]) => {
                                            return (
                                                <td>
                                                    <TruncatedString
                                                        toggled={item.Id === currentlyOpenedTd.id && key === currentlyOpenedTd.key}
                                                        toggleTd={() => toggleTd(item.Id, key)}
                                                        string={value}
                                                    />
                                                </td>
                                            );
                                        })}

                                        <td>
                                            <button>
                                                Edit
                                            </button>
                                        </td>
                                    </tr>
                                )
                            })
                        ) : (
                            <tr>
                                <td colSpan={tableHead.length}>
                                    <div
                                        className={'fetch-info'}
                                    >
                                        Not found
                                    </div>
                                </td>
                            </tr>
                        )
                    }
                    </tbody>
                </table>
                <div
                    className={'table-nav--bottom'}
                >
                    <TableNavigation
                        setPage={setCurrentPage}
                        currentPage={currentPage}
                        pages={pages}
                    />
                </div>
            </React.Fragment>
        );
    }
}

export default Table;
