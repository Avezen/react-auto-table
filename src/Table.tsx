import React from 'react';
import {TableHead} from "./TableHead";
import {TableBody} from "./TableBody";
import {TableColumnWidths} from "./TableColumnWidths";

export interface TableProps {
  data?: any[];
  sortBy: string;
  setSortBy: any;
  sortDirection: number;
  isFetching: boolean;
  error?: boolean;
  colWidths?: number[];
  tableHead: string[];
  currentlyOpenedTd: {
    id: null,
    key: null,
  };
  toggleTd: any;
  config: any;
}

export const Table = ({
  data,
  sortBy,
  setSortBy,
  sortDirection,
  isFetching,
  error,
  colWidths,
  tableHead,
  currentlyOpenedTd,
  toggleTd,
  config
}: TableProps) => (
    <table>
      <TableColumnWidths
          colWidths={colWidths}
      />
      <TableHead
          sortBy={sortBy}
          setSortBy={setSortBy}
          sortDirection={sortDirection}
          tableHead={tableHead}
          config={config}
      />
      <TableBody
          data={data}
          isFetching={isFetching}
          tableHead={tableHead}
          colWidths={colWidths}
          currentlyOpenedTd={currentlyOpenedTd}
          toggleTd={toggleTd}
          config={config}
      />
    </table>
);
