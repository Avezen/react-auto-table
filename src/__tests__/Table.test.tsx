import React from "react";
import {Table, TableProps} from "../Table";
import renderer from 'react-test-renderer';

function renderTable(props: Partial<TableProps> = {}) {
    const defaultProps: TableProps = {
        data: [
            {
                Id: 1,
                Name: 'Test 1'
            },
            {
                Id: 2,
                Name: 'Test 2'
            }
        ],
        sortBy: 'Id',
        setSortBy() {
            return;
        },
        sortDirection: 1,
        isFetching: false,
        error: false,
        colWidths: [50, 25],
        tableHead: ['Id', 'Name'],
        currentlyOpenedTd: {
            id: null,
            key: null,
        },
        toggleTd() {
            return;
        }
    };

    return <Table {...defaultProps} {...props} />;
}

describe("<Table />", () => {
    test("should render component with data, with tableHead, with colWidths, sorted ASC by Id", async () => {
        const findByTestId = renderTable(),
            TableComponent = renderer.create(findByTestId).toJSON();

        expect(TableComponent).toMatchSnapshot();
    });

    test("should render component with data, with tableHead, with colWidths, sorted DESC by Id", async () => {
        const props = {
            sortBy: 'Id',
            sortDirection: -1,
        };
        const findByTestId = renderTable(props),
            TableComponent = renderer.create(findByTestId).toJSON();

        expect(TableComponent).toMatchSnapshot();
    });

    test("should render component with data, with tableHead, with colWidths, sorted ASC by Name", async () => {
        const props = {
            sortBy: 'Name',
            sortDirection: 1,
        };
        const findByTestId = renderTable(props),
            TableComponent = renderer.create(findByTestId).toJSON();

        expect(TableComponent).toMatchSnapshot();
    });
});
