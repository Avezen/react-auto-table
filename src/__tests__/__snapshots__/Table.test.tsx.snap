// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render component with data, with tableHead, with colWidths, sorted ASC by Id 1`] = `
<table>
  <colgroup>
    <col
      width={50}
    />
    <col
      width={25}
    />
  </colgroup>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Id
        <i
          className="arrow down"
        />
      </th>
      <th
        onClick={[Function]}
      >
        Name
      </th>
      <th>
        Actions
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <button>
          Edit
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <button>
          Edit
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table /> should render component with data, with tableHead, with colWidths, sorted ASC by Name 1`] = `
<table>
  <colgroup>
    <col
      width={50}
    />
    <col
      width={25}
    />
  </colgroup>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Id
      </th>
      <th
        onClick={[Function]}
      >
        Name
        <i
          className="arrow down"
        />
      </th>
      <th>
        Actions
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <button>
          Edit
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <button>
          Edit
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Table /> should render component with data, with tableHead, with colWidths, sorted DESC by Id 1`] = `
<table>
  <colgroup>
    <col
      width={50}
    />
    <col
      width={25}
    />
  </colgroup>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Id
        <i
          className="arrow up"
        />
      </th>
      <th
        onClick={[Function]}
      >
        Name
      </th>
      <th>
        Actions
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <button>
          Edit
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <button>
          Edit
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
